src/coprocessor/endpoint.rs

```diff
@@ -11,641 +11,448 @@
+            .flatten()
+            .or_else(|e| Ok(make_error_response(e)))
     }
-}
 
-impl Display for RequestTask {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        f.debug_struct("RequestTask")
-            .field("req_ctx", &self.tracker.req_ctx)
-            .finish()
+    #[inline]
+    pub fn parse_and_handle_unary_request(
+        &self,
+        req: coppb::Request,
+        peer: Option<String>,
+    ) -> impl Future<Item = coppb::Response, Error = ()> {
+        let (handler_builder, req_ctx) = self.parse_request(req, peer, false);
+        self.handle_unary_request(req_ctx, handler_builder)
     }
-}
 
-impl<E: Engine> Runnable<Task<E>> for Host<E> {
-    // TODO: limit pending reqs
-    fn run(&mut self, _: Task<E>) {
-        panic!("Shouldn't call Host::run directly");
-    }
-
-    #[cfg_attr(feature = "cargo-clippy", allow(for_kv_map))]
-    fn run_batch(&mut self, tasks: &mut Vec<Task<E>>) {
-        let mut grouped_reqs = map![];
-        for task in tasks.drain(..) {
-            match task {
-                Task::Request(mut req) => {
-                    req.set_max_handle_duration(self.request_max_handle_duration);
-                    if let Err(e) = req.check_outdated() {
-                        let resp = err_resp(e, &mut self.basic_local_metrics);
-                        req.on_resp.respond(resp);
-                        continue;
+    /// The real implementation of handling a stream request.
+    ///
+    /// It first retrieves a snapshot, then builds the `RequestHandler` over the snapshot and
+    /// the given `handler_builder`. Finally, it calls the stream request interface of the
+    /// `RequestHandler` multiple times to process the request and produce multiple results.
+    // TODO: Convert to use async / await.
+    fn handle_stream_request_impl(
+        engine: E,
+        tracker: Box<Tracker>,
+        handler_builder: RequestHandlerBuilder<E::Snap>,
+    ) -> impl Stream<Item = coppb::Response, Error = Error> {
+        // When this function is being executed, it may be queued for a long time, so that
+        // deadline may exceed.
+
+        let tracker_and_handler_future = future::result(
+            tracker.req_ctx.deadline.check_if_exceeded(),
+        ).and_then(move |_| {
+            Self::async_snapshot(engine, &tracker.req_ctx.context)
+                .map(|snapshot| (tracker, snapshot))
+        })
+            .and_then(move |(tracker, snapshot)| {
+                // When snapshot is retrieved, deadline may exceed.
+                future::result(tracker.req_ctx.deadline.check_if_exceeded())
+                    .map(|_| (tracker, snapshot))
+            })
+            .and_then(move |(tracker, snapshot)| {
+                future::result(handler_builder.call_box((snapshot, &tracker.req_ctx)))
+                    .map(|handler| (tracker, handler))
+            });
+
+        stream::futures_ordered(vec![tracker_and_handler_future])
```

Comment:
Why not `tracker_and_handler_future.map(...).flatten_stream`?

Reply:
might not be applicable. tracker_and_handler_future is a future (of pair `(tracker, handler)`), not a future of a stream